rcl/src/rcl/type_version_hash.c

```diff
@@ -0,0 +1,257 @@
+// Copyright 2023 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
+#include <yaml.h>
+
+#include "rcl/allocator.h"
+#include "rcl/error_handling.h"
+#include "rcl/type_version_hash.h"
+#include "rcutils/types/char_array.h"
+#include "rcutils/sha256.h"
+#include "type_description_interfaces/msg/type_description.h"
+
+
+#define FAILFAST(expr) if (!(expr)) { \
+    rcl_set_error_state(emitter->problem, __FILE__, __LINE__); \
+    return 0; \
+}
+
+#define EMIT() FAILFAST(yaml_emitter_emit(emitter, &event))
+
+#define EMIT_KEY(key) \
+  FAILFAST( \
+    yaml_scalar_event_initialize( \
+      &event, NULL, NULL, \
+      (yaml_char_t *)key, strlen(key), \
+      0, 1, YAML_DOUBLE_QUOTED_SCALAR_STYLE)); \
+  EMIT();
+
+#define EMIT_STR_VAL(data, size) \
+  FAILFAST( \
+    yaml_scalar_event_initialize( \
+      &event, NULL, NULL, \
+      (yaml_char_t *)data, size, \
+      0, 1, YAML_DOUBLE_QUOTED_SCALAR_STYLE)); \
+  EMIT();
+
+#define EMIT_INT_VAL(value, fmt) { \
+    char decimal_buf[21]; \
+    FAILFAST( \
+      (size_t)snprintf(decimal_buf, sizeof(decimal_buf), fmt, value) < sizeof(decimal_buf)); \
+    FAILFAST( \
+      yaml_scalar_event_initialize( \
+        &event, NULL, NULL, \
+        (yaml_char_t *)decimal_buf, strlen(decimal_buf), \
+        1, 0, YAML_PLAIN_SCALAR_STYLE)); \
+    EMIT(); \
+}
```

Comment:
I have to say I'm not a huge fan of all of these macros. I know this is how we've done it in some other places, but macros like this just make reading and debugging this stuff hard later on. It might be good to figure out if we can avoid at least some of the macros here. But I don't have a great alternative, so this isn't a blocking comment from me.

Reply:
Part of me is perversely disappointed to note that my updated version has actually shortened the file length by one line. It seems this is a better pattern, let me know what you think of it now.